README.md
# Adaptable - the most adaptable moodle theme

Adaptable is a highly customisable responsive two column moodle theme based on the popular BCU theme adding:

* Customisable fonts (Google Fonts)
* Fully customisable colors
* Fully customisable block styles (including icons)
* Fully customisable buttons
* Additional header navigation
* News Ticker
* Alternative jQuery slider
* Alternative front page course styles
* Additional marketing blocks
* Additional layout settings for width, slider width, padding of various elements
* Social icons
* Mobile settings (customise how theme looks on mobile devices)
* Dismissible bootstrap alerts (can be made persistent by specifying a “key” which is stored in user preferences)
* Option to add login form in header on front page
* Favicon uploader

In addition many fields (menus, news items, alerts and help links) can be targeted using custom profile fields,
thus it is possible to present different users with different nagivation items and notices.
It is also possible for individual users to customise where they want top menu navigation to appear
(disable, home pages only, sitewide) using custom profile fields.

Adaptable has a lot of settings and may seem daunting at first, our advice is to simply install with the default
settings and play with it afterwards.

With a little time you should be able to setup an attractive Moodle site with a high degree of
individuality without without knowing any CSS.

This theme has been developed by:
* Jeremy Hopkins (Coventry University)
* Fernando Acedo (3-bits.com)


## HTML/CSS sample code for block areas

Here you will find some code samples to help you to customize the Info Box and the Marketing Blocks.

You can insert any HTML tag to customize the frontpage blocks. Use a ```<div>``` as a main container and add the height to keep the same value in all the blocks.

The Font Awesome icons set is available in
http://fortawesome.github.io/Font-Awesome/icons/.

You can insert any of them following the examples http://fortawesome.github.io/Font-Awesome/examples/


## Front Page Slider Styles

There are two possible slider styles each with different markup required:

### Original BCU Slider Markup:
```html
<div class="span9">
  <h4>Information</h4>
    Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore
    et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip
    ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
  </div>

  <div class="span3">
  <a href="#" class="submit">2013/14 Courses <i class="fa-chevron-right fa"></i></a>
</div>
```

### Coventry Style Slider Markup
```html
<div class="span6 col-sm-6">
<h3>Hand-crafted</h3> <h4>pixels and code for the Moodle community</h4>
<a href="#" class="submit">Check out our custom theme pricing!</a>
</div>
```

### Frontpage Marketing Block HTML structure Coventry
```html
<div><img src="http://somewebsite.com/2.jpg" class="marketimage"></div>
<h4><a href="#">International Courses</a></h4>
<p>Some text below the link....</p>
```


## Front page Info Box and Marketing Blocks

There are two Info blocks in the frontpage located above and below the Marketing Blocks. These are just for compatibility with the old BCU.

It is recommended to use the new marketing blocks builder that allows you to create your own layout and add much more blocks.

There are 8 rows where you can add up to 4 blocks in each. See pix/layout.png for more information.

You can enter any HTML code to the block, include FA icons, images, videos and apply inline styles.

Some samples:


#### Block with solid background, FA icon and some text:
```html
<div style="text-align:center; background: #e6e6e6; height: 240px; padding: 7px;">
    <i class="fa fa-paint-brush fa-5x" style="color: #009688;"></i>
    <h3>Title </h3>
    <div style="text-align:center;">Add your text here.</div>
</div>
```

#### Block with border and transparent background:
```html
<div style="text-align:center; height: 240px; padding: 7px; border: 1px solid #009688;">
    <i class="fa fa-list fa-5x" style="color: #009688;"></i>
    <h3>Heading</h3>
    <div style="text-align:center; padding: 5px; color: #009688;">Add your text here.</div>
</div>
```

#### Block with an image:
```html
<div style="height: 350px;">
    <img src="http://yoursite/yourimage.jpg" style="vertical-align:text-bottom; margin: 0 .5em;" height="auto" width="100%">
    <p style="margin-top: 5px; color: #333333; text-align: center;"><strong>Add your text here</strong></p>
</div>
```


#### Block with a video:
```html
<div style="background: #606060; height: 315px">
    <center>
    <iframe src="https://www.youtube.com/embed/wop3FMhoLGs" allowfullscreen="" frameborder="0" height="315" width="560"></iframe>
    </center>
</div>
```

#### Block using multi-lang filter:
```html
<div style="width: 100%; height: 240px; background-color: #cccccc;">
<h1 style="text-align: center; line-height: 120px;">
      <span class="multilang" lang="en">text in english</span>
      <span class="multilang" lang="es">texto en español</span>
      <span class="multilang" lang="fr">texte en français</span>
      <span class="multilang" lang="ca">text en català</span>
</h1>
</div>
```


## Footer Blocks

You can apply the same HTML/CSS in the footer blocks.

Some samples:

#### Contact information
```html
<i class="fa fa-building"></i>  High St. 100<br>
<span style="margin-left: 20px;">123456 City</span><br><br>
<i class="fa fa-phone"></i> +12 (3)456 78 90<br>
<i class="fa fa-envelope"></i> info@mail.com<br>
<i class="fa fa-globe"></i> www.example.com
```

#### List with Chevron
```html
<ul class="block-list">
    <li><a href="http://moodle.org/"><span class="fa fa-chevron-right"></span><span>Accessibility</span></a></li>
    <li><a href="http://moodle.org/"><span class="fa fa-chevron-right"></span><span>Moodle Help</span></a></li>
    <li><a href="http://moodle.org/"><span class="fa fa-chevron-right"></span><span>Moodle Feedback</span></a></li>
    <li><a href="http://moodle.org/"><span class="fa fa-chevron-right"></span><span>IT Help</span></a></li>
    <li><a href="http://moodle.org/"><span class="fa fa-chevron-right"></span><span>IT Feedback</span></a></li>
</ul>
```


## Copyright text

A sample of copyright text using FA icon
```html
Made with <i class="fa fa-heart" style="color: #ff0000;"></i> in Europe
```

## News Ticker

To add news to the ticker you need to create an unordered list using the Atto editor.

**IMPORTANT!** Verify you only enter the ```<li>``` tags. For example:
```html
<li>Configure all the theme colours</li>
<li>Use any Google Font for the content, headings and site title</li>
<li>Display a logo or a configurable title site</li>
<li>Configurable Slideshow</li>
<li>Display up to 12 marketing blocks in the front page</li>
```

## Versioning

Adaptable is maintained under the Semantic Versioning guidelines as much as possible.
Releases will be numbered with the following format:

**major**.minor.*patch*

and following these guidelines:
* Breaking backward compatibility bumps the major (and resets the minor and patch)
* New additions without breaking backward compatibility bumps the minor (and resets the patch)
* Bug fixes and misc changes bumps the patch

For more information on SemVer, please visit http://semver.org.


## Acknowledgments

Big thanks to all the volunteers that are collaborating and testing Adaptable continously.
We really appreciate your help and support to develope the most adaptable theme for moodle.


#### Development:
* Leonid Chernyavskiy
* COMETE (Paris Ouest University)
* Marina Glancy

#### Testing:
* Andrew Walding
* Alexander Goryntsev



## Licenses

Adaptable is licensed under the GPL v3 (GNU General Public License)
* http://www.gnu.org/licenses

Google Fonts released under:
* SIL Open Font License v1.1 - http://scripts.sil.org/OFL
* Apache 2 license - https://www.apache.org/licenses/LICENSE-2.0
* The Ubuntu fonts use the Ubuntu Font License v1.0  - http://font.ubuntu.com/ufl/ubuntu-font-licence-1.0.txt

The Font Awesome font (by Dave Gandy
* http://fontawesome.io) is licensed under the SIL OFL 1.1 - http://scripts.sil.org/OFL

Font Awesome CSS, LESS, and SASS files are licensed under the MIT License
* https://opensource.org/licenses/mit-license.html

Emojione (http://emojione.com) released under Creative Commons Attribution 4.0 International
* https://creativecommons.org/licenses/by/4.0/*


## Follow Us

Twitter: https://twitter.com/adaptable_theme
Facebook: https://www.facebook.com/adaptable.theme


### Modify it! - Improve it! - Share it!


CHANGES.md
Changes in version 1.2.5 (2016)
---------------------------------------
New features:
- Avatar beside user name in the navbar



Fixed issues:







Changes in version 1.2.4.2 (2016102300)
---------------------------------------
Warning: This is the last minor release compatible with moodle 2.9

Fixed issues:
- #527: Reset course and login blocks width 
- #526: Left padding in main-region 
- #521: Breadcrumbs do not display last item 
- #517: Student pictures narrowed in some listings
- #516: Notice: Undefined property: stdClass: 
- #515: Slider type 1 display caption background when empty 

Also includes some improvements in the mobile header but still working to fix some issues



Changes in version 1.2.4.1 (2016100300)
---------------------------------------
Warning: This is the last minor release compatible with moodle 2.9

Fixed issues:
- #512: Header social bug
- #511: Persisting horizontal scrollbar
- #509: Site title is not displayed without logo image


Changes in version 1.2.4 (2016092200)
-------------------------------------
Warning: This is the last minor release compatible with moodle 2.9

New features:
- Updated Google Fonts list
- Setting: Text selection colours
- Setting: Emoticons size 
- Setting: Piwik Analytics - Thanks to COMETE (Paris Ouest University)
- Setting: Navbar Menu hover colour

Fixed issues:
- #508: Integrate Analytics in one page 
- #507: Emoticon size 
- #504: Remove hardcoded strings in settings 
- #501: Fail to embed a video into marketing blocks 
- #500: Cannot get beyond New settings - Header on fresh install of moodle 3.1 with Adaptable in the theme directory 
- #499: User menu error in user ID in "My Blog" item 
- #497: Categories list on Front Page 
- #494: Definition of from35to100percent broken 
- #493: Inheriting from Adaptable: Hardcoded settings 
- #492: Textbox for email digest type in forums is too narrow 
- #491: Extraneous </div> in frontpage 
- #490: Heading / logo alignment 
- #489: Default titling does nothing without image 
- #487: Full login field label doesn't fit into its box and is cut from the end 
- #486: Logo is not responsive 
- #485: Option to have Search bar and Social Icons? 
- #484: Username placeholder uses not localized string for hint 
- #482: Untidy layout of the login screen 
- #481: Editor can't upload images in some settings sections 
- #480: Disable login button 
- #479: Mess with the front page layout when blocks are on the LEFT side 
- #478: Distorted front page layout before login 
- #476: Block region builder settings doesn't appear 
- #475: Navbar disappears when moving away from first page after login 
- #474: Convert 'course' string to theme string 
- #473: Add logo to default site title 
- #472: Site Title Maximum Width values up to 100% 
- #471: Docking blocks with transparent background 
- #419: Navbar caching "this course" list at site level


Changes in version 1.2.3 (2016052900)
-------------------------------------
- #469: Infobox2 strings missed
- #468: Lower Header blocks text and link Colour: not working
- #467: Override .block settings
- #444: Neither site title nor course title in the header


Changes in version 1.2.2 (2016050900)
-------------------------------------
- #449: 'People' item in 'This course' menu is not displayed in local language 
- #446: Failed to control 'Course Full Name' / 'Course Short Name' setting for the breadcrumb 
- #445: Header height has a 'hidden' limit 
- #444: Neither site title nor course title in the header 
- #441: Multi-lang content not displayed in the marketing blocks 
- #440: Preferences menu item enable by default 
- #429: Teacher Style front Page Tiles 


Changes in version 1.2.1 (2016031800)
-------------------------------------

- Fix #409 Bug rendering sub menus on mobile
- Fix #416 Bug rendering My Courses with cohort option selected
- Fix #411 Multiple tools menus added to menu
- Fix #406 User Preference setting doesn't apply
- Fix #414 Correct URL for My Feedback Report
- Fix #415 Analytics not rendering script tag properly
- Fix #413 Add setting for messages "badge" in header
- Fix #407 Fix time calculation on navbar cache ttl
- Fix #404 Alerts not displaying if set to display to logged in users
- Fix #398 Second infobox has not full screen
- Fix #278 Navbar sticky at the top
- Fix #394 Set Enable Home redirect=0 to On by default
- Fix #397 Replace <b> with <strong>
- Fix #395 Set site title to truncate and add width setting making consistent with course title

Changes in version 1.2.0 (2016030700)
-------------------------------------

- Fix #393 Fix HTML Validation errors on closing menu div
- Fix #392 Remove unwanted </i> tag from top menus
- Fix #389 Coursebox non closed <a> tag breaks HTML validation
- Fix #390 Header menu font-size increment is too rough
- Fix #375 Prevent Frontpage Block Regions appearing in block settings sitewide
- Fix #354 Grader Report: Icons in webkit overlapping.
- Fix #381 Prevent slider being wrapped in <a tag where no slider URL is given
- Fix #384 Add settings page for course layout
- Fix #184 Topic title header width changes when editing On
- Fix #383 Correct display of section headings in courses

lib.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Version details
 *
 * @package    theme_adaptable
 * @copyright  2015-2016 Jeremy Hopkins (Coventry University)
 * @copyright  2015-2016 Fernando Acedo (3-bits.com)
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 *
 */

define('THEME_ADAPTABLE_DEFAULT_ALERTCOUNT', '1');
define('THEME_ADAPTABLE_DEFAULT_ANALYTICSCOUNT', '1');
define('THEME_ADAPTABLE_DEFAULT_TOPMENUSCOUNT', '1');
define('THEME_ADAPTABLE_DEFAULT_TOOLSMENUSCOUNT', '1');
define('THEME_ADAPTABLE_DEFAULT_NEWSTICKERCOUNT', '1');
define('THEME_ADAPTABLE_DEFAULT_SLIDERCOUNT', '3');

/**
 * Parses CSS before it is cached.
 *
 * This function can make alterations and replace patterns within the CSS.
 *
 * @param string $css The CSS
 * @param theme_config $theme The theme config object.
 * @return string The parsed CSS The parsed CSS.
 */
function theme_adaptable_process_css($css, $theme) {

    // Set custom CSS.
    if (!empty($theme->settings->customcss)) {
        $customcss = $theme->settings->customcss;
    } else {
        $customcss = null;
    }
    $css = theme_adaptable_set_customcss($css, $customcss);

    // Define the default settings for the theme incase they've not been set.
    $defaults = array(
        '[[setting:linkcolor]]' => '#009688',
        '[[setting:linkhover]]' => '#009688',
        '[[setting:maincolor]]' => '#009688',
        '[[setting:backcolor]]' => '#FFFFFF',
        '[[setting:regionmaincolor]]' => '#FFFFFF',
        '[[setting:rendereroverlaycolor]]' => '#009688',
        '[[setting:rendereroverlayfontcolor]]' => '#FFFFFF',
        '[[setting:buttoncolor]]' => '#009688',
        '[[setting:buttontextcolor]]' => '#ffffff',
        '[[setting:buttonhovercolor]]' => '#00695c',
        '[[setting:buttonlogincolor]]' => '#ef5350',
        '[[setting:buttonloginhovercolor]]' => '#e53935',
        '[[setting:buttonlogintextcolor]]' => '#0084c2',
        '[[setting:buttonloginpadding]]' => '0px',
        '[[setting:buttonloginheight]]' => '24px',
        '[[setting:buttonloginmargintop]]' => '2px',
        '[[setting:buttonradius]]' => '5px',
        '[[setting:dividingline]]' => '#ffffff',
        '[[setting:dividingline2]]' => '#ffffff',
        '[[setting:navbarborder]]' => '#B7B3EF',
        '[[setting:navbarhover]]' => '#3C469C',
        '[[setting:breadcrumb]]' => '#b4bbbf',
        '[[setting:breadcrumbtextcolor]]' => '#444444',
        '[[setting:activebreadcrumb]]' => '#f5f5f5',
        '[[setting:breadcrumbseparator]]' => 'angle-right',
        '[[setting:loadingcolor]]' => '#f44336',
        '[[setting:footerbkcolor]]' => '#424242',
        '[[setting:footertextcolor]]' => '#ffffff',
        '[[setting:footertextcolor2]]' => '#ffffff',
        '[[setting:footerlinkcolor]]' => '#ffffff',
        '[[setting:headerbkcolor]]' => '#00796B',
        '[[setting:headerbkcolor2]]' => '#009688',
        '[[setting:headertextcolor]]' => '#ffffff',
        '[[setting:headertextcolor2]]' => '#ffffff',
        '[[setting:msgbadgecolor]]' => '#E53935',
        '[[setting:blockheadercolor]]' => '#009688',
        '[[setting:blockbackgroundcolor]]' => '#FFFFFF',
        '[[setting:blockheaderbackgroundcolor]]' => '#FFFFFF',
        '[[setting:blockbordercolor]]' => '#59585D',
        '[[setting:blockregionbackgroundcolor]]' => 'transparent',
        '[[setting:selectiontext]]' => '#000000',
        '[[setting:selectionbackground]]' => '#eeff41',
        '[[setting:marketblockbordercolor]]' => '#e8eaeb',
        '[[setting:marketblocksbackgroundcolor]]' => 'transparent',
        '[[setting:blocklinkcolor]]' => '#009688',
        '[[setting:blocklinkhovercolor]]' => '#009688',
        '[[setting:blockheaderbordertop]]' => '1px',
        '[[setting:blockheaderborderleft]]' => '0px',
        '[[setting:blockheaderborderright]]' => '0px',
        '[[setting:blockheaderborderbottom]]' => '0px',
        '[[setting:blockmainbordertop]]' => '0px',
        '[[setting:blockmainborderleft]]' => '0px',
        '[[setting:blockmainborderright]]' => '0px',
        '[[setting:blockmainborderbottom]]' => '0px',
        '[[setting:blockheaderbordertopstyle]]' => 'dashed',
        '[[setting:blockmainbordertopstyle]]' => 'solid',
        '[[setting:blockheadertopradius]]' => '0px',
        '[[setting:blockheaderbottomradius]]' => '0px',
        '[[setting:blockmaintopradius]]' => '0px',
        '[[setting:blockmainbottomradius]]' => '0px',
        '[[setting:coursesectionheaderbg]]' => '#F3F3F3',
        '[[setting:coursesectionheaderbordercolor]]' => '#F3F3F3',
        '[[setting:coursesectionheaderborderstyle]]' => '',
        '[[setting:coursesectionheaderborderwidth]]' => '',
        '[[setting:coursesectionheaderborderradiustop]]' => '',
        '[[setting:coursesectionheaderborderradiusbottom]]' => '',
        '[[setting:coursesectionborderstyle]]' => '1px',
        '[[setting:coursesectionborderwidth]]' => '',
        '[[setting:coursesectionbordercolor]]' => '#e8eaeb',
        '[[setting:coursesectionborderradius]]' => '',
        '[[setting:tilesbordercolor]]' => '#009688',
        '[[setting:slidermargintop]]' => '20px',
        '[[setting:slidermarginbottom]]' => '20px',
        '[[setting:currentcolor]]' => '#d9edf7',
        '[[setting:sectionheadingcolor]]' => '#5f588a',
        '[[setting:menufontsize]]' => '14px',
        '[[setting:menufontpadding]]' => '20px',
        '[[setting:topmenufontsize]]' => '14px',
        '[[setting:menubkcolor]]' => '#ffffff',
        '[[setting:menufontcolor]]' => '#444444',
        '[[setting:menuhovercolor]]' => '#b2dfdb',
        '[[setting:menubordercolor]]' => '#80cbc4',
        '[[setting:mobilemenubkcolor]]' => '#F9F9F9',
        '[[setting:mobilemenufontcolor]]' => '#000000',
        '[[setting:covbkcolor]]' => '#009688',
        '[[setting:covfontcolor]]' => '#ffffff',
        '[[setting:editonbk]]' => '#4caf50',
        '[[setting:editoffbk]]' => '#f44336',
        '[[setting:editverticalpadding]]' => '',
        '[[setting:edithorizontalpadding]]' => '',
        '[[setting:edittopmargin]]' => '',
        '[[setting:editfont]]' => '#ffffff',
        '[[setting:sliderh3color]]' => '#ffffff',
        '[[setting:sliderh4color]]' => '#ffffff',
        '[[setting:slidersubmitbgcolor]]' => '#009688',
        '[[setting:slidersubmitcolor]]' => '#ffffff',
        '[[setting:slider2h3color]]' => '#000000',
        '[[setting:slider2h4color]]' => '#000000',
        '[[setting:slider2h3bgcolor]]' => '#000000',
        '[[setting:slider2h4bgcolor]]' => '#ffffff',
        '[[setting:slideroption2color]]' => '#0066cc',
        '[[setting:slideroption2submitcolor]]' => '#ffffff',
        '[[setting:slideroption2a]]' => '#0066cc',
        '[[setting:socialsize]]' => '32px',
        '[[setting:socialsizemobile]]' => '22px',
        '[[setting:mobile]]' => '22',
        '[[setting:socialpaddingtop]]' => '1.8%',
        '[[setting:fontname]]' => 'Open Sans',
        '[[setting:fontsize]]' => '95%',
        '[[setting:fontheadername]]' => 'Roboto Slab',
        '[[setting:fontcolor]]' => '#333333',
        '[[setting:fontheadercolor]]' => '#333333',
        '[[setting:fontweight]]' => '400',
        '[[setting:fontheaderweight]]' => '400',
        '[[setting:fonttitlename]]' => 'Audiowide',
        '[[setting:fonttitleweight]]' => '400',
        '[[setting:fonttitlesize]]' => '48px',
        '[[setting:fonttitlecolor]]' => '#FFFFFF',
        '[[setting:fonttitlenamecourse]]' => 'Audiowide',
        '[[setting:fonttitleweightcourse]]' => '48px',
        '[[setting:fonttitlesizecourse]]' => '48px',
        '[[setting:fonttitlecolorcourse]]' => '#FFFFFF',
        '[[setting:sitetitlepaddingtop]]' => '0px',
        '[[setting:sitetitlepaddingleft]]' => '0px',
        '[[setting:coursetitlepaddingtop]]' => '0px',
        '[[setting:coursetitlepaddingleft]]' => '0px',
        '[[setting:searchboxpaddingtop]]' => '16px',
        '[[setting:pageheaderheight]]' => '72px',
        '[[setting:emoticonsize]]' => '16px',
        '[[setting:fullscreenwidth]]' => '98%',
        '[[setting:coursetitlemaxwidth]]' => '50%',
        '[[setting:sitetitlemaxwidth]]' => '50%',
        '[[setting:coursetitlepaddingtop]]' => '0px',
        '[[setting:hidebreadcrumbmobile]]' => '',
        '[[setting:hidepagefootermobile]]' => '',
        '[[setting:hidealertsmobile]]' => '',
        '[[setting:hidesocialmobile]]' => '',
        '[[setting:socialboxpaddingtopmobile]]' => '',
        '[[setting:socialboxpaddingbottommobile]]' => '',
        '[[setting:hidecoursetitlemobile]]' => '',
        '[[setting:hidelogomobile]]' => '',
        '[[setting:hideheadermobile]]' => '',
        '[[setting:enableheading]]' => 'fullname',
        '[[setting:breadcrumbtitle]]' => 'shortname',
        '[[setting:enableavailablecourses]]' => 'display',
        '[[setting:enableticker]]' => true,
        '[[setting:enabletickermy]]' => true,
        '[[setting:tickerwidth]]' => '',
        '[[setting:socialwallbackgroundcolor]]' => '#FFFFFF',
        '[[setting:socialwallsectionradius]]' => '6px',
        '[[setting:socialwallbordertopstyle]]' => 'solid',
        '[[setting:socialwallborderwidth]]' => '2px',
        '[[setting:socialwallbordercolor]]' => '#009688',
        '[[setting:socialwallactionlinkcolor]]' => '#009688',
        '[[setting:socialwallactionlinkhovercolor]]' => '#009688',
    );

    // Get all the defined settings for the theme and replace defaults.
    foreach ($theme->settings as $key => $val) {
        if (array_key_exists('[[setting:'.$key.']]', $defaults) && !empty($val)) {
            $defaults['[[setting:'.$key.']]'] = $val;
        }
    }

    $homebkg = '';
    if (!empty($theme->settings->homebk)) {
        $homebkg = $theme->setting_file_url('homebk', 'homebk');
        $homebkg = 'background-image: url("' . $homebkg . '");';
    }
    $defaults['[[setting:homebkg]]'] = $homebkg;

    // Replace the CSS with values from the $defaults array.
    $css = strtr($css, $defaults);
    if (empty($theme->settings->tilesshowallcontacts) || $theme->settings->tilesshowallcontacts == false) {
        $css = theme_adaptable_set_tilesshowallcontacts($css, false);
    } else {
        $css = theme_adaptable_set_tilesshowallcontacts($css, true);
    }
    return $css;
}


/**
 * Adds any custom CSS to the CSS before it is cached.
 *
 * @param string $css The original CSS.
 * @param string $customcss The custom CSS to add.
 * @return string The CSS which now contains our custom CSS.
 */
function theme_adaptable_set_customcss($css, $customcss) {
    $tag = '[[setting:customcss]]';
    $replacement = $customcss;
    if (is_null($replacement)) {
        $replacement = '';
    }

    $css = str_replace($tag, $replacement, $css);

    return $css;
}

/**
 * Set display of course contacts on front page tiles
 * @param string $css
 * @param string $display
 * @return $string
 */
function theme_adaptable_set_tilesshowallcontacts($css, $display) {
    $tag = '[[setting:tilesshowallcontacts]]';
    if ($display) {
        $replacement = 'block';
    } else {
        $replacement = 'none';
    }
    $css = str_replace($tag, $replacement, $css);
    return $css;
}

/**
 * Set user prefernces for zoom (show / hide block) function
 * @param moodle_page $page
 * @return void
 */
function theme_adaptable_initialise_zoom(moodle_page $page) {
    user_preference_allow_ajax_update('theme_adaptable_zoom', PARAM_TEXT);
    $page->requires->yui_module('moodle-theme_adaptable-zoom', 'M.theme_adaptable.zoom.init', array());
}

/**
 * Get the user preference for the zoom (show / hide block) function.
 */
function theme_adaptable_get_zoom() {
    return get_user_preferences('theme_adaptable_zoom', '');
}

/**
 * Set the user preference for full screen
 * @param moodle_page $page
 * @return void
 */
function theme_adaptable_initialise_full(moodle_page $page) {
    user_preference_allow_ajax_update('theme_adaptable_full', PARAM_TEXT);
    $page->requires->yui_module('moodle-theme_adaptable-full', 'M.theme_adaptable.full.init', array());
}

/**
 * Get the user preference for the zoom function.
 */
function theme_adaptable_get_full() {
    return get_user_preferences('theme_adaptable_full', '');
}

/**
 * Get the key of the last closed alert for a specific alert index.
 * This will be used in the renderer to decide whether to include the alert or not
 * @param int $alertindex
 */
function theme_adaptable_get_alertkey($alertindex) {
    user_preference_allow_ajax_update('theme_adaptable_alertkey' . $alertindex, PARAM_TEXT);
    return get_user_preferences('theme_adaptable_alertkey' . $alertindex, '');
}

/**
 * Get user preference for side block
 * Note: does not seem to be called / used within theme, may need to be removed
 */
function theme_adaptable_get_block_side() {
    static $theme;
    if (empty($theme)) {
        $theme = theme_config::load('adaptable');
    }
    return get_user_preferences('theme_adaptable_block_side', $theme->settings->blockside);
}

/**
 * Get HTML for settings
 * @param renderer_base $output
 * @param moodle_page $page
 */
function theme_adaptable_get_html_for_settings(renderer_base $output, moodle_page $page) {
    global $CFG;
    $return = new stdClass;

    $return->navbarclass = '';
    if (!empty($page->theme->settings->invert)) {
        $return->navbarclass .= ' navbar-inverse';
    }

    if (!empty($page->theme->settings->logo)) {
        $return->heading = html_writer::link($CFG->wwwroot, '', array('title' => get_string('home'), 'class' => 'logo'));
    } else {
        $return->heading = $output->page_heading();
    }

    $return->footnote = '';
    if (!empty($page->theme->settings->footnote)) {
        $return->footnote = '<div class="footnote">'.$page->theme->settings->footnote.'</div>';
    }

    return $return;
}

/**
 * Get theme setting
 * @param string $setting
 * @param string $format = false
 */
function theme_adaptable_get_setting($setting, $format = false) {
    static $theme;
    if (empty($theme)) {
        $theme = theme_config::load('adaptable');
    }

    if (empty($theme->settings->$setting)) {
        return false;
    } else if (!$format) {
        return $theme->settings->$setting;
    } else if ($format === 'format_text') {
        return format_text($theme->settings->$setting, FORMAT_PLAIN);
    } else if ($format === 'format_html') {
        return format_text($theme->settings->$setting, FORMAT_HTML, array('trusted' => true));
    } else {
        return format_string($theme->settings->$setting);
    }
}

/**
 * Serves any files associated with the theme settings.
 *
 * @param stdClass $course
 * @param stdClass $cm
 * @param context $context
 * @param string $filearea
 * @param array $args
 * @param bool $forcedownload
 * @param array $options
 * @return bool
 */
function theme_adaptable_pluginfile($course, $cm, $context, $filearea, $args, $forcedownload, array $options = array()) {
    static $theme;
    if (empty($theme)) {
        $theme = theme_config::load('adaptable');
    }
    if ($context->contextlevel == CONTEXT_SYSTEM) {
        if ($filearea === 'logo') {
            return $theme->setting_file_serve('logo', $args, $forcedownload, $options);
        } else if ($filearea === 'style') {
            theme_essential_serve_css($args[1]);
        } else if ($filearea === 'homebk') {
            return $theme->setting_file_serve('homebk', $args, $forcedownload, $options);
        } else if ($filearea === 'pagebackground') {
            return $theme->setting_file_serve('pagebackground', $args, $forcedownload, $options);
        } else if (preg_match("/p[1-9][0-9]/", $filearea) !== false) {
            return $theme->setting_file_serve($filearea, $args, $forcedownload, $options);
        } else if ((substr($filearea, 0, 9) === 'marketing') && (substr($filearea, 10, 5) === 'image')) {
            return $theme->setting_file_serve($filearea, $args, $forcedownload, $options);
        } else if ($filearea === 'iphoneicon') {
            return $theme->setting_file_serve('iphoneicon', $args, $forcedownload, $options);
        } else if ($filearea === 'iphoneretinaicon') {
            return $theme->setting_file_serve('iphoneretinaicon', $args, $forcedownload, $options);
        } else if ($filearea === 'ipadicon') {
            return $theme->setting_file_serve('ipadicon', $args, $forcedownload, $options);
        } else if ($filearea === 'ipadretinaicon') {
            return $theme->setting_file_serve('ipadretinaicon', $args, $forcedownload, $options);
        } else if ($filearea === 'fontfilettfheading') {
            return $theme->setting_file_serve('fontfilettfheading', $args, $forcedownload, $options);
        } else if ($filearea === 'fontfilettfbody') {
            return $theme->setting_file_serve('fontfilettfbody', $args, $forcedownload, $options);
        } else if ($filearea === 'adaptablemarkettingimages') {
            return $theme->setting_file_serve('adaptablemarkettingimages', $args, $forcedownload, $options);
        } else {
            send_file_not_found();
        }
    } else {
        send_file_not_found();
    }
}

/**
 * Get course activities for this course menu
 */
function theme_adaptable_get_course_activities() {
    GLOBAL $CFG, $PAGE, $OUTPUT;
    // A copy of block_activity_modules.
    $course = $PAGE->course;
    $content = new stdClass();
    $modinfo = get_fast_modinfo($course);
    $modfullnames = array();

    $archetypes = array();

    foreach ($modinfo->cms as $cm) {
        // Exclude activities which are not visible or have no link (=label).
        if (!$cm->uservisible or !$cm->has_view()) {
            continue;
        }
        if (array_key_exists($cm->modname, $modfullnames)) {
            continue;
        }
        if (!array_key_exists($cm->modname, $archetypes)) {
            $archetypes[$cm->modname] = plugin_supports('mod', $cm->modname, FEATURE_MOD_ARCHETYPE, MOD_ARCHETYPE_OTHER);
        }
        if ($archetypes[$cm->modname] == MOD_ARCHETYPE_RESOURCE) {
            if (!array_key_exists('resources', $modfullnames)) {
                $modfullnames['resources'] = get_string('resources');
            }
        } else {
            $modfullnames[$cm->modname] = $cm->modplural;
        }
    }
    core_collator::asort($modfullnames);

    return $modfullnames;
}

/**
 * Get formatted performance info showing only page load time
 * @param string $param
 */
function theme_adaptable_performance_output($param) {
    $html = html_writer::tag('span', get_string('loadtime', 'theme_adaptable').' '. round($param['realtime'], 2) . ' ' .
            get_string('seconds'), array('id' => 'load'));
    return $html;
}

/**
 * Initialize page
 * @param moodle_page $page
 */
function theme_adaptable_page_init(moodle_page $page) {
    global $CFG;
    $page->requires->jquery();
    // REMOVED: Deprecated function ->   error_log($CFG->version); ???
    if ($CFG->version < 2015051100) {
        $page->requires->jquery_plugin('bootstrap', 'theme_adaptable');
        $page->requires->jquery_plugin('dropdown', 'theme_adaptable');
    }

    $page->requires->jquery_plugin('pace', 'theme_adaptable');
    $page->requires->jquery_plugin('flexslider', 'theme_adaptable');
    $page->requires->jquery_plugin('ticker', 'theme_adaptable');
    $page->requires->jquery_plugin('easing', 'theme_adaptable');
    $page->requires->jquery_plugin('adaptable', 'theme_adaptable');
}

/**
 * Strip full site title from header
 * @param string $heading
 */
function theme_adaptable_remove_site_fullname($heading) {
    global $SITE, $PAGE;
    if (strpos($PAGE->pagetype, 'course-view-') === 0) {
        return $heading;
    }

    $header = preg_replace("/^".$SITE->fullname."/", "", $heading);

    return $header;
}
